include/boost/math/statistics/univariate_statistics.hpp

```diff
@@ -12,75 +12,101 @@
 #include <cmath>
 #include <vector>
 #include <boost/assert.hpp>
+#include <atomic>
+#include <type_traits>
+#include <execution>
+#include <numeric>
+#include <valarray>
 
 namespace boost::math::statistics {
 
-template<class ForwardIterator>
-auto mean(ForwardIterator first, ForwardIterator last)
+template<class ExecutionPolicy, class ForwardIterator>
+auto mean(ExecutionPolicy&& exec, ForwardIterator first, ForwardIterator last)
 {
     using Real = typename std::iterator_traits<ForwardIterator>::value_type;
     BOOST_ASSERT_MSG(first != last, "At least one sample is required to compute the mean.");
-    if constexpr (std::is_integral<Real>::value)
+    if constexpr (std::is_integral_v<Real>)
     {
-        double mu = 0;
-        double i = 1;
-        for(auto it = first; it != last; ++it) {
-            mu = mu + (*it - mu)/i;
-            i += 1;
+        if constexpr (std::is_same_v<std::remove_reference_t<decltype(exec)>, decltype(std::execution::seq)>)
+        {
+            double mu = 0;
+            double i = 1;
+            for(auto it = first; it != last; ++it) {
+                mu = mu + (*it - mu)/i;
+                i += 1;
+            }
+            return mu;
+        }
+        else
+        {
+            return std::reduce(exec, first, last, 0.0) / std::distance(first, last);
```

Comment:
Take a read the Higham's "Accuracy and Stability of Numerical Algorithms", equation 1.6a,b to see why I did it this way and not with `std::reduce`.

Another good reference is:

Robert F Ling. Comparison of several algorithms for computing sample means and variances. Journal of the American Statistical Association, 69(348): 859–866, 1974

Reply:
What about for integer arithmetic? From [26.7.1](http://www.open-std.org/Jtc1/sc22/wg21/docs/papers/2016/p0024r1) ```reduce applies binary_op in an unspecified order, which yields a non-deterministic result for non-associative or non-commutative binary_op such as floating-point addition.``` I will have to look at those references for the other cases.

Reply:
Yeah, for integer arithmetic this won't work; then again the result needs to be cast to a floating point type anyway.

If indeed we wished to support integer arithmetic then we'd need to figure out whether we want to cast every element to double and use the floating point algorithm or use the naive algorithm and then cast the sum to double. Overflow is the danger of the latter.